### PR TITLE
overview: re-implenent shutdown/restart with split button

### LINKED ISSUE #13316: *Restart* button and kebab menu that replaced drop-down menu is unclear
Cockpit version: 209
OS: Fedora
Page: Front

Prior to Cockpit 209, the 'Power options' menu looked like this:

powerdrop

In Cockpit 209, it looks like this:

powerkebab

I understand the problem that we're trying to solve here, but I find the new design unclear.

Per the Patternfly docs, the previous element is a [standard dropdown menu](https://www.patternfly.org/v4/documentation/react/components/dropdown). The new approach is listed as ['with kebab'](https://www.patternfly.org/v4/documentation/react/components/dropdown#with-kebab) and is the eighth option in the list. (The name 'kebab menu' was apparently coined in [this tweet](https://twitter.com/lukew/status/590911111358242816?lang=en), which is pretty disparaging...)

It appears the 'kebab menu' element was introduced as part of Google's material design, so it's worth looking at [Google's docs](https://material.io/components/menus/#dropdown-menu) here. (Neither the Patternfly and Bootstrap docs really cover what these elements are for, that I can find, though there's a [Patternfly blog post from 2016](https://blog.patternfly.org/communication/an-eventstask-drawer-thats-handy/) which talks about kebabs a bit.)

In the Google docs examples of kebab menus they're all quite clearly related by design to the UI element they relate to, but in Cockpit 209 the kebab button is not clearly associated with the **Restart** button next to it except by proximity. They're not part of a single clearly-identifiable UI element. The button is a button, and the kebab menu is just sort of floating in space next to it. It wasn't really clear to me that the kebab menu is related to the button, that it gives you other possible actions besides **Restart** (the default choice).

Note in the Google docs that the kebab menu is used to present a set of *possible actions that can be taken on a given element* - in particular, look at the example under **Behaviour**, where we have several items in a list, each with a kebab menu offering 'Preview', 'Share', 'Get Link' etc. These are not 'either or' choices, but actions. They're not alternatives to the list item - they're things you can do *to it*.

Now if you read down to [*Exposed dropdown menu*](https://material.io/components/menus/#exposed-dropdown-menu) and you'll read this:

"Exposed dropdown menus display the currently selected menu item above the menu. They can be used only when a single menu item can be chosen at a time."

above a picture of exactly the type of dropdown Cockpit *used* to use here. So, thinking about this, I can kinda see the problem the kebab button is trying to fix. The idea of a dropdown is that you have one state selected, you click it and select another state, and the button's state (and the actual effect in the app) is updated. That's not *quite* what we're doing here; we actually have a list of possible actions and we want the chosen action to take effect immediately when clicked (IIRC that's what happens, anyway). So I can see why someone thought a dropdown wasn't quite the right thing.

However I don't think a kebab floating in space next to a button is quite right either, looking at all the docs above and other discussions of menu styles around the internet. We seem to be sort of awkwardly between the two scenarios here. I'm afraid I don't have a great solution to propose, but I just wanted to flag up that I found this a bit confusing and provide a bit of research and ask if it can be reconsidered.

Thanks!

### PR DESCRIPTION
**Pull request by KKoukiou**
Fixes #13316

### CODE REVIEW

**Comment by KKoukiou** [comment]
@garrett you will see that the caret button is a bit misassligned. At first I thought i was again some PF-4 inline style not getting applied but not really. I tried it as a standalone component outside of cockpit and it works fine which means we are overwriting something.

**Comment by martinpitt** [comment]
Needs a [test adjustment](https://209.132.184.41:8493/logs/pull-13394-20200114-152121-1def7f78-cockpit-project-cockpit--fedora-31/log.html#67-2)

**Comment by martinpitt** [comment]
@KKoukiou : We also got rhel-8-2-distropkg now, so this at least temporarily will need another special case. I plan to land this one in 8.2 still, so they'll go away soon again. Sorry for the extra paperwork :/

**Comment by KKoukiou** [comment]
@garrett check this: 
Screen Shot 2020-01-15 at 11 49 38

**Comment by garrett** [comment]
It looks fine in Chrome, but not Firefox.

For some reason, Chrome has _inline CSS_ that adjusts the alignment, whereas Firefox does not.

Regardless of browser, we shouldn't have inline CSS anyway, and this CSS _should_ be included in stylesheets if it's really necessary.

(FWIW: I think it's a workaround to how the icon was created.)

**TL;DR**: This looks like yet-another case of PF4 using inline styles. :roll_eyes:

**Comment by KKoukiou** [comment]
> It looks fine in Chrome, but not Firefox.
> 
> For some reason, Chrome has _inline CSS_ that adjusts the alignment, whereas Firefox does not.
> 
> Regardless of browser, we shouldn't have inline CSS anyway, and this CSS _should_ be included in stylesheets if it's really necessary.
> 
> (FWIW: I think it's a workaround to how the icon was created.)
> 
> **TL;DR**: This looks like yet-another case of PF4 using inline styles. roll_eyes

Initially I thought so as well, but them built a tiny app with only presenting that split button with disabled the inline styles as well, and it works fine.
You can build it easily from this repo here https://github.com/KKoukiou/patternfly-react-table-example/tree/split-button

And in that I don't see any inline styles in the dev tools.

**Comment by garrett** [comment]
It's inline in the dev tools of Chrome. (Look at the highlighted line.)

svg-inline

Here's the same in Firefox. (See how there's no inline style?):

svg-inline-ff

Does that tiny app have CSP to disable inline CSS?

**Comment by KKoukiou** [comment]
> It's inline in the dev tools of Chrome. (Look at the highlighted line.)
> 
> svg-inline
> 
> Emphasis on the `style="vertical-align: -0.125em;"`.
> 
> Here's the same in Firefox. (See how there's no inline style?):
> 
> svg-inline-ff
> 
> Does that tiny app have CSP to disable inline CSS?

Above I said that even though I don't see inline styles in Firefox when testing my tiny app it works as expected. And sure - I have CSP to disable inline CSS there as well.

**Comment by garrett** [comment]
PF4's button alignment depends on two things:
- specific font metrics
- inline CSS that bumps down the icons (in some cases, like in Chrome)

To change it to proper center alignment, regardless of font, we'd need CSS like the following:

```css
.pf-c-dropdown__toggle {
  align-items: stretch;
  /* ...Or the align-items: center; can just be removed from the original CSS, as stretch is the default for align-items. */
}

.pf-c-dropdown__toggle-button {
  display: flex;
  align-items: center;
}
```

**Comment by garrett** [comment]
I've applied the CSS from my previous comment in `lib/page.css` and tested it in Firefox, Chrome, and Epiphany.

I also was going to test in Edge... but...

win10

I'm sure it works there. :crossed_fingers:

**Comment by garrett** [comment]
I just reinstalled a fresh copy of Windows in a new VM to test Edge. It works as expected; the alignment is perfect there too.

File: test/verify/check-login [hunk 1]
```diff
@@ -379,9 +379,13 @@ account    required     pam_succeed_if.so user ingroup %s""" % m.get_admin_group
             b.wait_present("#shutdown-group button[data-action=restart][data-stable=yes]")
             b.wait_present("#shutdown-group button.disabled")
             b.wait_not_present("#shutdown-group button:not(.disabled)")
-        else:
+        elif m.image == "rhel-8-2-distropkg":
             b.wait_present("#restart-button[disabled][data-stable=yes]")
             b.wait_present("#shutdown-group.disabled[data-stable=yes]")
+        else:
+            b.wait_present("#restart-button[disabled]")
+            b.wait_present("#shutdown-group[disabled]")
```
Comment (martinpitt) [inline test/verify/check-login]: Please put back the data-stable. Your changed code still sets it, and it avoids the test matching on the `null` state where it's not yet known whether the user is privileged or not.

**Comment by martinpitt** [comment]
Hmm, I immediately get a crash with this:
```
Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.

Check the render method of `OverviewPage`. 2 overview.js:223962:21
    createFiberFromTypeAndProps http://localhost:9090/cockpit/@localhost/system/overview.js:223962
    createFiberFromElement http://localhost:9090/cockpit/@localhost/system/overview.js:223985
    createChild http://localhost:9090/cockpit/@localhost/system/overview.js:212291
    reconcileChildrenArray http://localhost:9090/cockpit/@localhost/system/overview.js:212565
    reconcileChildFibers http://localhost:9090/cockpit/@localhost/system/overview.js:212931
    reconcileChildren http://localhost:9090/cockpit/@localhost/system/overview.js:215422
    updateFragment http://localhost:9090/cockpit/@localhost/system/overview.js:215623
    beginWork$1 http://localhost:9090/cockpit/@localhost/system/overview.js:217548
    callCallback http://localhost:9090/cockpit/@localhost/system/overview.js:197669
    invokeGuardedCallbackDev http://localhost:9090/cockpit/@localhost/system/overview.js:197718
    invokeGuardedCallback http://localhost:9090/cockpit/@localhost/system/overview.js:197773
    beginWork$$1 http://localhost:9090/cockpit/@localhost/system/overview.js:223113
    performUnitOfWork http://localhost:9090/cockpit/@localhost/system/overview.js:222031
    workLoopSync http://localhost:9090/cockpit/@localhost/system/overview.js:222004
    performSyncWorkOnRoot http://localhost:9090/cockpit/@localhost/system/overview.js:221603
    scheduleUpdateOnFiber http://localhost:9090/cockpit/@localhost/system/overview.js:221031
    updateContainer http://localhost:9090/cockpit/@localhost/system/overview.js:224436
    legacyRenderSubtreeIntoContainer http://localhost:9090/cockpit/@localhost/system/overview.js:224861
    unbatchedUpdates http://localhost:9090/cockpit/@localhost/system/overview.js:221766
    legacyRenderSubtreeIntoContainer http://localhost:9090/cockpit/@localhost/system/overview.js:224860
    render http://localhost:9090/cockpit/@localhost/system/overview.js:224941
    init http://localhost:9090/cockpit/@localhost/system/overview.js:244833
```
